src/java.base/share/classes/java/security/PEMDecoder.java

```diff
@@ -0,0 +1,486 @@
+            };
+        } catch (GeneralSecurityException | IOException e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    /**
+     * Decodes and returns a {@link DEREncodable} from the given {@code String}.
+     *
+     * <p> This method reads the {@code String} until PEM data is found
+     * or the end of the {@code String} is reached.  If no PEM data is found,
+     * an {@code IllegalArgumentException} is thrown.
+     *
+     * <p> This method returns a Java API cryptographic object,
+     * such as a {@code PrivateKey}, if the PEM type is supported.
+     * Any non-PEM data preceding the PEM header is ignored by the decoder.
+     * If no cryptographic object is found, a {@link PEMRecord} will be
+     * returned containing the type identifier and Base64-encoded data.
+     * Any non-PEM data preceding the PEM header will be stored in
+     * {@code leadingData}.
+     *
+     * <p>If no PEM data is found, an {@code IllegalArgumentException} is
+     * thrown.
+     *
+     * <p> Input consumed by this method is read in as
+     * {@link java.nio.charset.StandardCharsets#UTF_8 UTF-8}.
+     *
+     * @param str a String containing PEM data
+     * @return a {@code DEREncodable}
+     * @throws IllegalArgumentException on error in decoding or no PEM data
+     * found
+     * @throws NullPointerException when {@code str} is null.
+     */
+    public DEREncodable decode(String str) {
+        Objects.requireNonNull(str);
+        DEREncodable de;
+        try {
+            return decode(new ByteArrayInputStream(
+                str.getBytes(StandardCharsets.UTF_8)));
+        } catch (IOException e) {
+            // With all data contained in the String, there are no IO ops.
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    /**
+     * Decodes and returns a {@link DEREncodable} from the given
+     * {@code InputStream}.
+     *
+     * <p>This method reads the {@code InputStream} until PEM data is
+     * found or until the end of the stream.  It can be called repeatedly
+     * on the {@code InputStream} to decode additional PEM data.
+     * The {@code InputStream} read pointer will remain at the end of
+     * the PEM footer unless an IOException occurs.
+     *
+     * <p> This method returns a Java API cryptographic object,
+     * such as a {@code PrivateKey}, if the PEM type is supported.
+     * Any non-PEM data preceding the PEM header is ignored by the decoder.
+     * If no cryptographic object is found, a {@link PEMRecord} will be
+     * returned containing the type identifier and Base64-encoded data.
+     * Any non-PEM data preceding the PEM header will be stored in
+     * {@code leadingData}.
+     *
+     * <p> If no PEM data is found, an {@code IllegalArgumentException} is
+     * thrown.
+     *
+     * @param is InputStream containing PEM data
+     * @return a {@code DEREncodable}
+     * @throws IOException on IO error with the {@code InputStream}
```

Comment:
Since IOE is not purely IO error and contains format errors that are not recoverable. Do you want to say more?